Report menu and camera fixes

ethankoester opened the pull request:
- Can’t report players without filling out required fields
- Dropdown open buttons no longer animate when not allowed
- Fixed weird tween effect with the “game or player” when switching to
  the report page
- Disabled selectors can no longer be used by clicking them in the
  center
- Camera zoom level persists between modes, preventing players from
  getting into third person when first person is forced

Commit:
Fix Report Menu and Camera

- Can’t report players without filling out required fields
- Dropdown open buttons no longer animate when not allowed
- Fixed weird tween effect with the “game or player” when switching to
the report page
- Disabled selectors can no longer be used by clicking them in the
center
- Camera zoom level persists between modes, preventing players from
getting into third person when first person is forced

Commit:
See previous commit

Didn’t realize they had to be selected to commit

hockeychaos commented on CoreScriptsRoot/Modules/Settings/Pages/ReportAbuseMenu.lua:
```diff
@@ -165,6 +165,45 @@ local function Initialize()
+			-- Returns: Bool active
+			-- Activates/deactivates the submit button based on current state
+
+			local active = areFieldsCompleted()
```
a little weird to have active be the name from areFieldsCompleted... maybe use fieldsCompleted?

hockeychaos commented on CoreScriptsRoot/Modules/Settings/Pages/ReportAbuseMenu.lua:
```diff
@@ -190,39 +229,41 @@ local function Initialize()
-					spawn(function()
-						game.Players:ReportAbuse(nil, abuseReason, this.AbuseDescription.Selection.Text)
-					end)
+			if areFieldsCompleted() then
```
lets switch this to the negative comparison and just return immediately, makes the code a little less messy

hockeychaos commented on CoreScriptsRoot/Modules/Settings/Utility.lua:
```diff
@@ -923,13 +921,15 @@ local function CreateSelector(selectionStringTable, startPosition)
-			this:SetSelectionIndex(newIndex)
-			if usesSelectedObject() then
-				GuiService.SelectedCoreObject = this.SelectorFrame
+			if interactable then
```
lets switch this to the negative comparison and just return immediately, makes the code a little less messy

Commit:
Mobile control improvements and cleanup

- Touch jump button disappears when humanoid jump power is 0 or when
humanoid jump is disabled
- Virtual thumbstick now appears at the correct position when the touch
is on the center of the frame
- Made requested cleanup to utility and report menu scripts

hockeychaos commented on PlayerScripts/PlayerScriptsPlace/StarterPlayer/StarterPlayerScripts/ControlScript.rbxmx:
```diff
@@ -337,10 +339,14 @@ UserInputService.GamepadConnected:connect(function(gamepadEnum)
 	
 	ControlModules.Gamepad:Enable()
 end)
+MasterControl.ControlStepped:connect(function() -- Must be done constantly because there is no Humanoid.StateEnabledChanged event
```
We will need to fix this before shipping, lets add a new event to humanoid for this

hockeychaos commented on PlayerScripts/PlayerScriptsPlace/StarterPlayer/StarterPlayerScripts/ControlScript/MasterControl/TouchJump.rbxmx:
```diff
@@ -31,11 +33,13 @@ local TOUCH_CONTROL_SHEET = "rbxasset://textures/ui/TouchControlsSheet.png"
 --[[ Public API ]]--
 function TouchJump:Enable()
 	JumpButton.Visible = true
+	TouchJump.Enabled = true
```
What does this variable do?

Commit:
Ethan2

Ethan2

Commit:
Merge pull request #272 from ROBLOX/Ethan-CustomButtons

Ethan2

Commit:
Fix button hover states

- Buttons now respond to hover states correctly

Commit:
Merge pull request #275 from ROBLOX/Ethan-CustomButtons

Fix button hover states

SolarCrane commented on CoreScriptsRoot/CoreScripts/ContextActionTouch.lua:
```diff
@@ -171,6 +208,65 @@ function createNewButton(actionName, functionInfoTable)
+
+	while not complete do
+		if buttonPositionsOccupied[index] then
+			index = index + 1
```
It would be cleaner to move out the index search to a separate function named some like FindOpenPosition()

SolarCrane commented on CoreScriptsRoot/CoreScripts/ContextActionTouch.lua:
```diff
@@ -171,6 +208,65 @@ function createNewButton(actionName, functionInfoTable)
+			)
+
+			position, occupyingPosition = pos, index
+			complete = true
```
We can get rid of this 'complete' variable if we move out the search logic mentioned above

SolarCrane commented on CoreScriptsRoot/Modules/BackpackScript.lua:
```diff
@@ -208,7 +229,7 @@ local function EquipNewTool(tool) --NOTE: HopperBin
 
 local function IsEquipped(tool)
-	return tool and ((tool:IsA('HopperBin') and tool.Active) or tool.Parent == Character) --NOTE: HopperBin
+	return not not (tool and ((tool:IsA('HopperBin') and tool.Active) or tool.Parent == Character)) --NOTE: HopperBin
```
This is a no-no, if you want to convert to a bool for nil object checks please use:
`local myBool = (object ~= nil)`

SolarCrane commented on CoreScriptsRoot/CoreScripts/ContextActionTouch.lua:
```diff
@@ -171,6 +208,65 @@ function createNewButton(actionName, functionInfoTable)
+						totalInArc = numInArc
+						break
+					end
+				end
```
The above arc calculation can be moved into its own function as well for increased readability.
Then we can replace the break with a return statement.
Also, if the number of buttons exceed all that can fit in the arcs does it always return position as 0?

SolarCrane commented on CoreScriptsRoot/Modules/BackpackScript.lua:
```diff
@@ -663,6 +719,150 @@ local function MakeSlot(parent, index)
+		end
+	end
+
+	-- Find closest currently-loaded tools that are below amd above the saved configuration
```
> amd 

This really isn't a big deal but looks like a typo

SolarCrane commented on CoreScriptsRoot/Modules/BackpackScript.lua:
```diff
@@ -663,6 +719,150 @@ local function MakeSlot(parent, index)
+
+			if otherSavedIndex ~= nil then
+				if otherSavedIndex < savedIndex then -- Other tool was saved below this tool
+					if closestBelow == nil or savedIndex - closestBelowIndex > savedIndex - otherSavedIndex then
```
> savedIndex - closestBelowIndex > savedIndex - otherSavedIndex

Is mathematically equivalent to 
`-closestBelowIndex > -otherSavedIndex`
which is equal to
`closestBelowIndex < otherSavedIndex`

SolarCrane commented on CoreScriptsRoot/Modules/BackpackScript.lua:
```diff
@@ -663,6 +719,150 @@ local function MakeSlot(parent, index)
+						end
+					end
+				elseif otherSavedIndex > savedIndex then -- Other tool was saved above this tool
+					if closestAbove == nil or closestAboveIndex - savedIndex > otherSavedIndex - savedIndex then
```
Same math logic as above

SolarCrane commented on CoreScriptsRoot/Modules/Settings/Pages/ReportAbuseMenu.lua:
```diff
@@ -165,6 +165,45 @@ local function Initialize()
 			submitText.ZIndex = 1
 		end
 
+		local function fieldsCompleted()
```
we should rename this to `areFieldsCompleted()` so that the function has a verb name.

SolarCrane commented on CoreScriptsRoot/Modules/Settings/Utility.lua:
```diff
@@ -313,6 +324,89 @@ local function isGuiVisible(gui, debug) -- true if any part of the gui is visibl
+		info.HoverState = hoverState
+
+		-- Connections
+		button.AncestryChanged:connect(evaluateStorage)
```
When do we clean up this AncestryChanged connection?

SolarCrane commented on CoreScriptsRoot/Modules/Settings/Utility.lua:
```diff
@@ -2017,4 +2127,7 @@ function moduleApiTable:TweenProperty(instance, prop, start, final, duration, ea
 end
 
+---------------------- Main code -----------------------
+RunService.RenderStepped:connect(DetectCustomButtonHoverStates)
```
This is a CRITICAL problem:
Utility is for various helper and math functions to cut-down on commonly re-used code. We cannot be connecting to global scope events here, especially ones like RenderStepped which happens at a high frequency.

SolarCrane commented on PlayerScripts/PlayerScriptsPlace/StarterPlayer/StarterPlayerScripts/ControlScript.rbxmx:
```diff
@@ -686,6 +717,10 @@ UserInputService.GamepadConnected:connect(function(gamepadEnum)
 end)
 
+MasterControl.ControlStepped:connect(function() -- Must be done constantly because there is no Humanoid.StateEnabledChanged event
+	ControlModules.Touch:EvaluateJumpButton()
```
Can you work with Ben on this one? Rather than connecting checks to run every step we could add a Changed event for you.

SolarCrane commented on CoreScriptsRoot/Modules/Settings/Utility.lua:
```diff
@@ -1655,7 +1764,7 @@ local function AddNewRow(pageToAddTo, rowDisplayName, selectionType, rowValues,
 
 	local RowFrame = nil
-	RowFrame = Util.Create'ImageButton'
+	RowFrame, RowFrameInfo = MakeCustomButton'ImageButton'
```
RowFrameInfo is in the global name scope since you didn't specify local for it. This might cause some unintended behaviors, so lets move it to be a local variable too.

BinaryResolved commented:
Should we just close this? It seems like it will just be here forever otherwise.
